midend/flattenLogMsg.cpp

```diff
@@ -0,0 +1,198 @@
+#include "midend/flattenLogMsg.h"
+#include "flattenLogMsg.h"
+
+namespace P4 {
+
+const IR::Type_StructLike* FindTypeInLogMsgForReplace::hasStructInParameter(
+        const IR::MethodCallStatement* methodCallStatement) {
+    if (auto* path = methodCallStatement->methodCall->method->to<IR::PathExpression>()) {
+        if (path->path->name.name == "log_msg") {
+            auto* param1 = methodCallStatement->methodCall->arguments->at(0)->expression;
+            auto* param2 = methodCallStatement->methodCall->arguments->at(1)->expression;
+            if (!param1->is<IR::StringLiteral>() || !param2->is<IR::StructExpression>()) {
+                return nullptr;
+            }
+            auto* type =
+                typeMap->getTypeType(methodCallStatement->methodCall->typeArguments->at(0), true);
+            if (auto* typeStruct = type->to<IR::Type_StructLike>()) {
+                for (auto field : typeStruct->fields) {
+                    if (field->type->is<IR::Type_StructLike>()) {
+                        return typeStruct;
+                    }
+                }
+            }
+        }
+    }
+    return nullptr;
+}
+
+const IR::MethodCallStatement* FindTypeInLogMsgForReplace::prepareLogMsgStatement(
+        const IR::MethodCallStatement* methodCallStatement) {
+    if (auto* newMethod = getReplacementMethodCall(methodCallStatement->id)) {
+        return newMethod;
+    }
+    // Create new statement
+    auto* param1 = methodCallStatement->methodCall->arguments->at(0)->expression;
+    std::string strParam1 = param1->to<IR::StringLiteral>()->value.c_str();
+    auto* argType =
```

Comment:
even here you have to check that the method has 1 type argument.

Reply:
This situation is impossible here, because this code executable if we have found corresponded MethodCallStatement for replacement in the map. The record in the map appears if log_smg has two arguments only.